casper/src/main/scala/coop/rchain/casper/blocks/BlockReceiver.scala

```diff
@@ -0,0 +1,280 @@
+import coop.rchain.shared.Log
+import coop.rchain.shared.syntax._
+import fs2.Stream
+import fs2.concurrent.Queue
+
+sealed trait RecvStatus
+// Checking if block should be ignored or continue with processing
+case object BeginReceived extends RecvStatus
+// Block received and stored in blocks store, waiting for validation and DAG insertion
+case object ReceivedAndStored extends RecvStatus
+// Block sent to validation
+case object PendingValidation extends RecvStatus
+// Requested missing dependencies
+case object PendingRequest extends RecvStatus
+
+object BlockReceiverState {
+  def apply[MId]: BlockReceiverState[MId] = new BlockReceiverState(
+    blocksSt = Map(),
+    receiveSt = Map(),
+    childRelations = Map()
+  )
+}
+
+final case class BlockReceiverState[MId](
+    /**
+      * Blocks received and stored in BlockStore with parent relations
+      */
+    blocksSt: Map[MId, Set[MId]],
+    /**
+      * Blocks receiving status
+      */
+    receiveSt: Map[MId, RecvStatus],
+    /**
+      * Blocks mapping with children relations
+      */
+    childRelations: Map[MId, Set[MId]]
+) {
+
+  /**
+    * Keep status of block as received process started
+    *  - thread safe sync like opening transaction for a block by block hash
+    */
+  def beginReceived(id: MId): (BlockReceiverState[MId], Boolean) = {
+    val expectedReceive = receiveSt.get(id).collect { case PendingRequest => true }.getOrElse(true)
+    // If state is not known or pending it's request, continue with receiving
+    if (expectedReceive) {
+      // Update state to begin received status
+      val newReceiveSt = receiveSt + ((id, BeginReceived))
+      (copy(receiveSt = newReceiveSt), true)
+    } else {
+      (this, false)
+    }
+  }
+
+  /**
+    * End of block receiving status update
+    *  - like closing transaction, block received and stored
+    *
+    *  @return unseen parent dependencies
+    */
+  def endReceived(id: MId, parents: List[(MId, Boolean)]): (BlockReceiverState[MId], Set[MId]) = {
+    val curStateOpt = receiveSt.get(id)
+    assert(
+      curStateOpt == BeginReceived.some,
+      s"Received should be called only in begin received state, actual: $curStateOpt, hash: $id"
+    )
+    curStateOpt
+      .collect {
+        case BeginReceived =>
+          // Update blocks state, keep unseen parents only
+          val parentsNotStored = parents.filter(_._2).map(_._1).toSet
+          val unseenParents    = parentsNotStored -- blocksSt.keySet -- receiveSt.keySet - id
+          val newBlocksSt      = blocksSt + ((id, unseenParents))
+
+          // Update block status to received and set unseen parents to Pending receive state
+          val newReceiveStored  = receiveSt + ((id, ReceivedAndStored))
+          val newPendingReceive = unseenParents.map((_, PendingRequest))
+          val newReceiveSt      = newReceiveStored ++ newPendingReceive
+
+          // Update children relations of received block
+          val parentsHashes = parents.map(_._1)
```

Comment:
We use `MId` as the block (message) identifier. And here we explicitly indicate that this is a hash. These are implementation details. Maybe rename to `parentsIds`?